internal/servers/controller/internal/metric/cluster.go

```diff
@@ -0,0 +1,177 @@
+	// to the boundary controller cluster is.
+	gRpcRequestSize prometheus.ObserverVec = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: clusterSubSystem,
+			Name:      "grpc_request_size_bytes",
+			Help:      "Histogram of request sizes for HTTP requests.",
+			// 100 bytes, 1kb, 10kb, 100kb, 1mb, 10mb, 100mb, 1gb
+			Buckets: gRpcMsgSizeBuckets,
+		},
+		[]string{labelGRpcCode, labelGRpcService, labelGRpcMethod},
+	)
+
+	// gRpcResponseSize collections measurements of how large each response
+	// from the boundary controller cluster is.
+	gRpcResponseSize prometheus.ObserverVec = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: clusterSubSystem,
+			Name:      "grpc_response_size_bytes",
+			Help:      "Histogram of response sizes for HTTP responses.",
+			// 100 bytes, 1kb, 10kb, 100kb, 1mb, 10mb, 100mb, 1gb
+			Buckets: gRpcMsgSizeBuckets,
+		},
+		[]string{labelGRpcCode, labelGRpcService, labelGRpcMethod},
+	)
+)
+
+func statusFromError(err error) *status.Status {
+	if s, ok := status.FromError(err); ok {
+		return s
+	}
+	return status.New(codes.Unknown, "Unknown Code")
+}
+
+func splitMethodName(fullMethodName string) (string, string) {
+	fullMethodName = strings.TrimPrefix(fullMethodName, "/") // remove leading slash
+	if i := strings.Index(fullMethodName, "/"); i >= 0 {
+		return fullMethodName[:i], fullMethodName[i+1:]
+	}
+	return "unknown", "unknown"
+}
+
+type requestRecorder struct {
+	labels prometheus.Labels
+
+	// measurements
+	reqSize *int
+	start   time.Time
+}
+
+func newRequestRecorder(ctx context.Context, req interface{}, fullMethodName string) requestRecorder {
+	const op = "metric.newRequestRecorder"
+	service, method := splitMethodName(fullMethodName)
+	r := requestRecorder{
+		labels: prometheus.Labels{
+			labelGRpcMethod:  method,
+			labelGRpcService: service,
+		},
+		start: time.Now(),
+	}
+
+	reqProto, ok := req.(proto.Message)
+	switch {
+	case ok:
+		reqSize := proto.Size(reqProto)
+		r.reqSize = &reqSize
+	default:
+		event.WriteError(ctx, op, errors.New(ctx, errors.Internal, op, "unable to cast to proto.Message"))
+	}
+	return r
+}
+
+func (r requestRecorder) record(ctx context.Context, resp interface{}, err error) {
+	const op = "metric.(requestRecorder).record"
+	st := statusFromError(err)
+	r.labels[labelGRpcCode] = st.Code().String()
+
+	gRpcRequestLatency.With(r.labels).Observe(time.Since(r.start).Seconds())
+
+	if r.reqSize != nil {
+		gRpcRequestSize.With(r.labels).Observe(float64(*r.reqSize))
+	}
+
+	if respProto, ok := resp.(proto.Message); ok {
+		respSize := proto.Size(respProto)
+		gRpcResponseSize.With(r.labels).Observe(float64(respSize))
+	} else {
+		event.WriteError(ctx, op, errors.New(ctx, errors.Internal, op, "unable to cast to proto.Message"))
+	}
+}
+
+// InstrumentClusterInterceptor wraps a UnaryServerInterceptor and measures
```

Comment:
I think this sentence might be accidentally cut short but I also thing it might not be, sorry if this is intentional 😅.

Reply:
Updated the comment to hopefully be more complete.  Thank you!